Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp

```diff
@@ -3969,6 +4113,61 @@ B3::PatchpointValue* B3IRGenerator::createTailCallPatchpoint(BasicBlock* block,
     return patchpoint;
 }
 
+bool B3IRGenerator::canInline() const
+{
+    unsigned inlineDepth = 0;
+    Checked<uint32_t> inlinedBytes = 0;
+    for (auto* currentFrame = this; currentFrame; currentFrame = currentFrame->m_inlineParent) {
+        ++inlineDepth;
+        inlinedBytes += currentFrame->m_inlinedBytes;
+    }
+    return inlineDepth < Options::maximumWasmDepthForInlining() && inlinedBytes.value() < Options::maximumWasmCallerSizeForInlining();
+}
+
+auto B3IRGenerator::emitInlineDirectCall(uint32_t calleeFunctionIndex, const TypeDefinition& calleeSignature, Vector<ExpressionType>& args, ResultList& resultList) -> PartialResult
+{
+    Vector<Value*> getArgs;
+
+    for (auto* arg : args)
+        getArgs.append(m_currentBlock->appendNew<VariableValue>(m_proc, B3::Get, origin(), arg));
+
+    BasicBlock* continuation = m_proc.addBlock();
+
+    const FunctionData& function = m_info.functions[calleeFunctionIndex];
+    m_protectedInlineeGenerators.constructAndAppend(*this, CompilationMode::OMGForInlineCallMode, calleeFunctionIndex, continuation, WTFMove(getArgs));
```

Comment:
Instead of creating `CompilationMode::OMGForInlineCallMode`, we should just pass the current CompilationMode, and separately adding a boolean flag etc. to indicate that this is inlined calleframe.